modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/PartitionListener.java

```diff
@@ -604,6 +604,12 @@ public boolean onBeforeApply(Command command) {
         return false;
     }
 
+    @Override
+    public void onLeaderStop() {
+        maxObservableSafeTime = -1;
```

Comment:
Is it a problem or not that we do not have any defence against concurrent `PartitionListener#onBeforeApply` and `PartitionListener#onLeaderStop`?

Reply:
That should not be a problem because before onBeforeApply we shoould verify whether we are on leader on not. Thus I don't expect onBeforeApply to be called after node lost it's leadership.

Reply:
I was thinking about the scenario when we've started to handle onBeforeApply, 
and in the place where we write `maxObservableSafeTime = proposedSafeTime`
```
            if (proposedSafeTime >= maxObservableSafeTime) {
                maxObservableSafeTime = proposedSafeTime;
            }
```
right before that `onLeaderStop` logic is performed, so we will have
 ```
maxObservableSafeTime == some value > 0
maxObservableSafeTime == -1
maxObservableSafeTime == proposedSafeTime
```
and finished `onLeaderStop` logic, so  `maxObservableSafeTime` won't be -1 as we were expected

Reply:
I'll create follow up ticket for that. The issue is urgent. Given solution makes TC much more stable.